Ovn bridge mappings api

maiqueb opened the pull request:
This PR introduces syntactic sugar for the OVN API to configure the bridge mappings through which OVN packets enter the underlay.

It addresses [this RFE](https://bugzilla.redhat.com/show_bug.cgi?id=2209690), which hopes to simplify (and make less error prone...) the configuration of the underlay for OVN-Kubernetes multi-networks.
More information about the OVN bridge mappings can be found [here](https://access.redhat.com/documentation/en-us/red_hat_openstack_platform/16.0/html/networking_guide/bridge-mappings).

cathay4t commented on rust/src/lib/ovs.rs:
```diff
@@ -201,12 +221,166 @@ impl MergedOvsDbGlobalConfig {
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
+#[serde(rename_all = "lowercase", deny_unknown_fields)]
+pub enum OvnBridgeMappingState {
```
Please include `#[non_exhaustive]` to ensure we don't break rust API when adding new state.

maiqueb replied:
Done.

cathay4t commented on rust/src/lib/ovsdb/db.rs:
```diff
@@ -421,3 +473,99 @@ impl OvsDbUpdate {
+    }
+}
+
+#[cfg(test)]
+mod tests {
```
Please move test code to folder `rust/src/lib/unit_tests`

maiqueb replied:
I had it there, but I honestly couldn't get this to pass the linters because of the rust modules. I'll need some help with that.

maiqueb replied:
Done.

cathay4t reviewed (changes requested):
If this syntax sugar, the plugins should not required to make any changes, all the wrapping should be done at top level `lib/ovs.rs` file. The `ovsdb/db.rs` do not need to handle this.

maiqueb commented:
> If this syntax sugar, the plugins should not required to make any changes, all the wrapping should be done at top level `lib/ovs.rs` file. The `ovsdb/db.rs` do not need to handle this.

OK, I'll take a look into that.

maiqueb committed:
API: add ovn subtree

This commit adds a new subtree - `ovn` - to the `NetworkState`. This
feature requires the ovsdb module enabled, since it is essentially just
syntactic sugar on top of the ovsdb `ovn-bridge-mappings` external ID
key.

To add a new bridge-mapping the user should apply the following snippet:
```yaml
ovn:
  bridge-mappings:
  - localnet: tenantblue
    bridge: ovsbr1
    state: present
```

This API supports partial editing; thus, to delete a particular mapping,
the user **must** set its state to absent:
```yaml
ovn:
  bridge-mappings:
  - localnet: tenantred
    bridge: ovsbr1
    state: absent
```

Keep in mind the `bridge` attribute is optional when deleting.

Signed-off-by: Miguel Duarte Barroso <mdbarroso@redhat.com>

maiqueb committed:
ovsdb: error when ovn-bridge-mappings ext id key is used

The users should use the proper API to manipulate the OVN bridge
mappings; to make it more explicit to the user, we throw an error when
they attempt to manipulate the external ID key directly.

Signed-off-by: Miguel Duarte Barroso <mdbarroso@redhat.com>

maiqueb committed:
ovsdb: preserve the `ovn-bridge-mappings` value

When the user clears the ovsdb external IDS, preserve the existing
bridge mappings. In case the user wants to remove these mappings from
the OVS system, they should use the upper level OVN API (which supports
partial editing) by setting the OVN bridge mapping's state to absent.

Signed-off-by: Miguel Duarte Barroso <mdbarroso@redhat.com>

cathay4t commented on rust/src/lib/ovs.rs:
```diff
@@ -176,7 +182,25 @@ impl MergedOvsDbGlobalConfig {
+                    };
                 } else {
+                    if ex_ids.get(OVN_BRIDGE_MAPPINGS).is_some() {
+                        const INVALID_EXTERNAL_IDS_KEY: &str = "Cannot use the `ovn-bridge-mappings` external_ids key directly. Please use ovs_db.ovn-bridge-mappings API instead";
```
Please change `ovs_db.ovn-bridge-mappings` to `ovn.bridge-mappings`.

Please also break long line with `\`

maiqueb replied:
Good catch.

maiqueb replied:
Done.

cathay4t commented on rust/src/lib/ovn.rs:
```diff
@@ -0,0 +1,268 @@
+        let v = serde_json::Value::deserialize(deserializer)?;
+
+        if let Some(v) = v.as_object() {
+            if let Some(v) = v.get("bridge-mappings") {
```
If you are doing `Deserialize` by yourself, you need to handle invalid key like:

```
ovn:
  invalid-set:
    abc: 1
  bridge-mappings:
    - localnet: yellow
      abc: ab
      bridge: ovsbr3
```

The `#[serde(deny_unknown_fields)]` can do the magic if using auto deserializer `#[derive(Deserialize)]`. From my view, you do not need special deserializer here.

maiqueb replied:
OK, I'll throw away the custom deserializer and run with derived.

maiqueb replied:
Still trying to do this ... Facing some issues.

maiqueb replied:
I'm thinking about this some more and there may be a reason to use a custom deserializer - although not a very good one.

If I use a custom deserializer I can provide a better error msg to the user if they try to provision the following bridge mapping:
```yaml
ovn:
  bridge-mappings:
  - localnet: net1
```

The above example is wrong because the user **must** also provide the bridge.

Right now, all I get is:
```
[2023-07-07T07:37:51Z INFO  nmstate::query_apply::net_state] Retrying on: VerificationError: Verification failure: ovn.bridge-mappings desire '[{"localnet":"net1"}]', current '[]'
```

Thoughts @cathay4t ?

cathay4t replied:
The `sanitize()` function is designed post-deserialize check and clean up.

maiqueb replied:
Added a `sanitize` func to ensure a mappings being added / modified has the required info.

Also moved asserting there aren't duplicate `localnet` keys in the desired state to a separate sanitize func.

Finally, removed the custom deserializer.

cathay4t commented on rust/src/lib/ovn.rs:
```diff
@@ -0,0 +1,268 @@
+                    }
+                    Err(err) => {
+                        return Err(serde::de::Error::custom(format!(
+                        "Error unmarshalling `OvnBridgeMapping` {v:?}: {err}"
```
Normally we do not scare people with debug output `{:?}`.

maiqueb replied:
Hm, I had just copied this from https://github.com/nmstate/nmstate/blob/f0f7c1edd8152dc03e6287c2222b524167c6f128/rust/src/lib/ovs.rs#L114

Well, doesn't matter, I'm trying to see if I can avoid using the custom deserializer - if so, all this code will be gone.

maiqueb replied:
This was removed, since the deserializer code is now gone.

cathay4t commented on rust/src/lib/ovn.rs:
```diff
@@ -0,0 +1,268 @@
+
+#[derive(Debug, Clone, PartialEq, Eq, Default, Serialize)]
+#[non_exhaustive]
+pub struct OvnConfiguration {
```
Public document required. Try `/// `

maiqueb replied:
Done - I pretty much just moved it from the `OvnBridgeMapping` struct to here. 

Let me know if we need more docs.

cathay4t commented on rust/src/lib/ovn.rs:
```diff
@@ -0,0 +1,268 @@
+#[non_exhaustive]
+pub enum OvnBridgeMappingState {
+    Present,
+    Absent,
```
In my test, this YAML is sufficient enough to delete a map:

```yml
ovn:
  bridge-mappings:
    - localnet: red
      state: absent
```

If is on purpose, please add public document for this usage.

maiqueb replied:
It is (enough to delete, and on purpose).

I'll add the public doc.

maiqueb replied:
Done.

cathay4t commented on tests/integration/ovs_test.py:
```diff
@@ -638,6 +645,69 @@ def test_remove_all_ovsdb_global_config():
     }
 
 
+def test_ovsdb_global_config_add_delete_mapping():
```
Please add two more test cases:
 1. Delete one map out of two+ maps.
 2. Replace/Modify a map via:
```yml
ovn:
  bridge-mappings:
    - localnet: yellow
      state: absent
    - localnet: yellow
      bridge: ovsbr5
```

maiqueb replied:
I think I see; your objective is to assert we can
1. remove without naming the bridge
2. state is optional when adding / preserving

Right ?

maiqueb replied:
Wait, I just now noticed your second example features the **same** localnet key for both listed mappings.

That's weird... Is that intentional ? I really don't know what the expected behavior should be.
@cathay4t

cathay4t replied:
> 1. Delete one map out of two+ maps.

This is to make sure nmstate will not touch undesired ovn bridge map.

> 2. Replace/Modify a map via

This is for changing OVN mapping in single YML.

maiqueb replied:
We've decided offline overriding the bridge would provide a good user experience, and thus we'll implement that.

To make the usage less error prone, we've also decided to reject desired states where the bridge mappings feature duplicate `localnet` keys.

Provided an e2e + unit test for this scenario.

I chose to just reject, instead of printing the exact duplicated localnet key, since that would require more code.

Let me know if you prefer me to take that route (i.e. point at the duplicate error key in the error msg).

cathay4t reviewed (changes requested):
In general, this patch looks very good and also passed all my manual tests.

Just some lints to fix on test coverage, error message, unknown key and documentation.

Thank you!

maiqueb committed:
ovn: do not allow for duplicate localnet keys

This commit makes it impossible for the user to provide multiple
`OvnBridgeMappings` with the same `localnet` key in the desired state.

The way for updating a key is to simply refer to it once, and provide
the update `bridge` value. When doing that, the `state` attribute is not
required. Refer to the following snippet for an example (assuming an
existing `net1: br1` mapping):

```yaml
ovn:
  bridge-mappings:
  - localnet: net1
    bridge: new-bridge-123
```

The example above will change the `net1` `localnet` mapping to the
`new-bridge-123` OVS bridge.

Signed-off-by: Miguel Duarte Barroso <mdbarroso@redhat.com>

maiqueb commented:
failing on one lane on an unrelated test: `test_new_bond_uses_mac_of_first_port_by_name`.

Can we somehow request that particular lane to be re-executed ? @cathay4t / @ffmancera

cathay4t commented:
> failing on one lane on an unrelated test: `test_new_bond_uses_mac_of_first_port_by_name`.

It is a known random failure never able to fix or even reproduced in my VM.

> 
> Can we somehow request that particular lane to be re-executed ? @cathay4t / @ffmancera

Restarted

maiqueb committed:
ovn: ensure add/modify mappings require bridge

Adds a sanitize function that ensures each mapping being provisioned -
or updated - provides the bridge over which the localnet traffic will be
forwarded.

Signed-off-by: Miguel Duarte Barroso <mdbarroso@redhat.com>

cathay4t commented on rust/src/lib/net_state.rs:
```diff
@@ -99,6 +102,13 @@ pub struct NetworkState {
     pub ovsdb: OvsDbGlobalConfig,
+    #[serde(
+        default,
+        rename = "ovn",
```
No need to rename.

maiqueb replied:
Wow of course :sweat_smile: 

Done.

cathay4t commented on rust/src/lib/query_apply/net_state.rs:
```diff
@@ -309,6 +312,41 @@ impl NetworkState {
+
+                                let mut sorted_mappings =
+                                    updated_mappings.clone();
+                                sorted_mappings.sort_by(|v1, v2| {
```
`sort_unstable_by()` is quicker.

maiqueb replied:
Done.

cathay4t reviewed (approved):
Good enough to merge. Just two lints:
 * No need to rename from `ovn` to `ovn` in serializer.
 * Use `sort_unstable_by()` which is quicker(according to rust doc).
